pkg/agent/openflow/pipeline.go

```diff
@@ -624,6 +632,112 @@ func (c *client) localProbeFlow(localGatewayIP net.IP, category cookie.Category)
 		Done()
 }
 
+func (c *client) snatFlows(uplinkOfport uint32, bridgeLocalPort int, nodeIP net.IP, category cookie.Category) []binding.Flow {
+	snatIPRange := &binding.IPRange{nodeIP, nodeIP}
+	vMACInt, _ := strconv.ParseUint(strings.Replace(globalVirtualMAC.String(), ":", "", -1), 16, 64)
+	flows := []binding.Flow{
+		// Resubmit the packet from the uplink interface to conntrackTable.
+		c.pipeline[classifierTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchInPort(uplinkOfport).
+			Action().LoadRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
+			Action().ResubmitToTable(conntrackTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Enforce IP packet into the conntrack zone with SNAT. If the connection is SNATed, the reply packet should use
+		// Pod IP as the destination, and then resubmit to conntrackStateTable.
+		c.pipeline[conntrackTable].BuildFlow(priorityNormal).MatchProtocol(binding.ProtocolIP).
+			Action().CT(false, conntrackStateTable, ctZone).NAT().CTDone().
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Rewrite dMAC with the global vMAC if the packet is a reply to the Pod from the external address
+		c.pipeline[conntrackStateTable].BuildFlow(priorityHigh).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(false).MatchCTStateTrk(true).
+			MatchCTMark(snatCTMark).
+			MatchRegRange(int(marksReg), markTrafficFromUplink, binding.Range{0, 15}).
+			Action().LoadRange(binding.NxmFieldDstMAC, vMACInt, binding.Range{0, 47}).
+			Action().ResubmitToTable(dnatTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Resubmit the packet sent from local Pod to the external IP address to dnatTable.
+		c.pipeline[conntrackStateTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
+			MatchCTStateNew(false).MatchCTStateTrk(true).
+			MatchCTMark(snatCTMark).
+			Action().ResubmitToTable(dnatTable).
+			Cookie(c.cookieAllocator.Request(category).Raw()).
+			Done(),
+		// Output the non-SNAT packet to the bridge interface directly if it is received from the uplink interface.
+		c.pipeline[conntrackStateTable].BuildFlow(priorityNormal).
+			MatchProtocol(binding.ProtocolIP).
```

Comment:
For non-IP packets, we have already output to the bridge interface before entering conntrack(In latest code, it should be in uplink interface table, https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/openflow/pipeline.go#L483). Here we only focused on IP packets. It is because the OVS conntrack requires IP packets.

Reply:
Ok. Another question - I checked https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/openflow/pipeline.go#L1143. Does it mean we no longer support kube-proxy on Windows, since the flow only sets macRewriteMark but does not use vMAC?
And it does not support the traffic from Node IP to a remote Pod, because it matches only the local subnet?

Reply:
We still use kube-proxy on Windows for NodePort Service traffic. In my mind the `macRewriteMark` is introduced in AntreaProxy, it is because OVS DNAT only changes the dst IP, but not modifies dst MAC. We use this mark to enforce OVS could leverage the dst IP to calculate the dst MAC and target ofport number.

As for the flow you mentioned, it is for the reply packet of a local Pod accessing NodePort Service. Since the reply packet is sent out from the bridge interface (where the Node's IP is configured) and it uses the Node's gateway (not antrea-gw0) MAC as dst when it enters OVS, we need to rewrite the dst MAC with the client Pod's.

This flow should doesn't match the request packet from Node IP to a remote Pod, which still leverages the L3Routing flows to decide the destination. As for vMAC, it is used in L3Routing, we don't merge it with `macRewriteMark` yet.

Reply:
I assume you mean MAC rewrite by this flow: https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/openflow/pipeline.go#L660?
Then only when AntreaProxy is enabled, it checks macRewriteMark, otherwise it checks vMAC. That is why I ask does it mean we no more support kube-proxy on Windows.

Reply:
The flow to rewrite MAC is this one https://github.com/vmware-tanzu/antrea/blob/master/pkg/agent/openflow/pipeline.go#L683

kube-proxy is needed on Windows, we only cover ClusterIP Service in AntreaProxy(so for ClusterIP Service case, kube-pproxy is not needed). But for NodePort Service, we still leverage kube-proxy.